### DISCUSSION TITLE
Can source generator perform this

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#88972] [General]
Run the app for once, produce stuffs that's only producible after the app is running, and the stuffs are used for making release build more efficient.

For example, compiler might be unable to do some function fusion like turning 'a calls b calls c' to simply a call to a more efficient function, which I can produce by running my application and looking at the result

Usually there's a specification that was easy for me to build things, but specifications are complicated and I can flatten it out by computing the expression once and giving it a function that transform the result which is a function too to a more efficient function

**Comment by huoyaoyuan** [comment]
> turning 'a calls b calls c' to simply a call to a more efficient function

Sounds like the purpose of [interceptor](https://github.com/dotnet/csharplang/issues/7009).

**Reply by Xyncgas** [reply]
is interceptor capable for actually running the app to redirect to the optimized route

Because I can produce a optimized route only when the app runs, but I would like this optimized route to be compiled into my release build, so people get speedy version of the app not the version for developer / designers when it's running

Ideally, I can generate an optimized route after running my app, that will be cached for future compilations unless people change the source code and they have to regenerate the optimized version from the designer version of the codes they wrote using my library.

Further in detail, the designer version has people writing a bunch of transformer functions to objects, resulting in a final object that can alternatively be generated by a special function taking their 'delta' so to speak with some special vectors. So running the transformer functions would be very slow if there's a bunch of them, and as designers do their jobs there will be a bunch of them, but the optimized route is fast, today we are already producing the optimized route when the app is running, it just use the special function to generate the _product_ object in one call. It's only natural for wanting to make it faster do I wish to not generate the optimize route in release build when the app's running but as part of the compilation step where I can run my app first and produce some presidency for the optimized paths to go in the release build.

Hopefull the technology is there and I can already do this with the great tooling microsoft and you in the .NET community has provided

**Comment by KennethHoff** [comment]
If I understand you correctly, this sounds like something https://github.com/dotnet/csharplang/discussions/1028 could solve?

**Comment by BreyerW** [comment]
So that sounds like two _distinct_ steps and one is actual runtime run. Source generator cannot do that. The only way to make it work would be to have separate tool (eg. command-line console project) that generates necessary data for extra optimization in format that can be consumed by source generator - which is basically c# code only AFAIK (so it could be a single file with just assembly attribute that has string parameter or something that SG can look up and extract to drive generation). But honestly at this point you might as well do full source-gen with roslyn or other tool, skipping SG entirely unless you already are halfway there with SG
